chain/pool/txwithtime.go

```diff
@@ -0,0 +1,229 @@
+
+	v, ok := tp.twtMap.Delete(txnHash)
+	if ok {
+		tp.twtCount--
+		twt = v.(*txnWithTime)
+		txnHash := twt.txn.Hash()
+		tp.updateTwtFingerprint(txnHash)
+	}
+
+	return
+}
+
+// replace an oldtxn by a new txn in twtMap
+func (tp *TxnPool) replaceTwt(oldTxn, newTxn *transaction.Transaction) (twt *txnWithTime) {
+	if oldTxn == nil || newTxn == nil {
+		return
+	}
+
+	tp.removeTwt(oldTxn.Hash())
+	tp.appendTwt(newTxn)
+
+	return
+}
+
+// romve expired items from txn with time list
+func (tp *TxnPool) RemoveExpiredTwt() (expiredTwt []*txnWithTime) {
+
+	now := time.Now().UnixMilli()
+	expiredThreshold := now - MaxSyncTxnInterval
+
+	delKeys := make([]interface{}, 0)
+
+	tp.twtMapMu.Lock()
+	defer tp.twtMapMu.Unlock()
+
+	for pair := tp.twtMap.Oldest(); pair != nil; pair = pair.Next() {
+		twt := pair.Value.(*txnWithTime)
+		if twt.arriveTime < expiredThreshold { // expired
+			delKeys = append(delKeys, pair.Key)
+			expiredTwt = append(expiredTwt, pair.Value.(*txnWithTime))
+		} else {
+			break
+		}
+	}
+
+	for _, key := range delKeys {
+		tp.twtMap.Delete(key)
+		tp.twtCount--
+		tp.updateTwtFingerprint(key.(common.Uint256))
+	}
+
+	return
+}
+
+// remove submitted txns from txn with time map
+func (tp *TxnPool) removeFromTwtMap(txnsRemoved []*transaction.Transaction) {
+	var totalRemoved int = 0
+
+	tp.twtMapMu.Lock()
+	defer tp.twtMapMu.Unlock()
+
+	for _, txn := range txnsRemoved {
+		txnHash := txn.Hash()
+		_, ok := tp.twtMap.Load(txnHash)
+		if ok {
+			tp.twtMap.Delete(txnHash)
+			tp.updateTwtFingerprint(txnHash)
+			totalRemoved++
+		}
+	}
+	tp.twtCount -= totalRemoved
+
+}
+
+// get address and nonce from txn with time map
+func (tp *TxnPool) GetTwtAddrNonce() map[common.Uint160]uint64 {
+
+	mapAddrNonce := make(map[common.Uint160]uint64)
+
+	tp.twtMapMu.Lock()
+	defer tp.twtMapMu.Unlock()
+
+	for pair := tp.twtMap.Oldest(); pair != nil; pair = pair.Next() {
+
+		twt := pair.Value.(*txnWithTime)
+		sender, err := twt.txn.GetProgramHashes()
+		if err != nil {
+			return nil
+		}
+		nonce, _ := mapAddrNonce[sender[0]]
+		if twt.txn.UnsignedTx.Nonce+1 > nonce {
+			mapAddrNonce[sender[0]] = twt.txn.UnsignedTx.Nonce + 1 // we send expecected(next) nonce to neighbor
+		}
+	}
+
+	return mapAddrNonce
+}
+
+// get transactions later than specific time from txn with time map
+// earliest is in milli-second
+func (tp *TxnPool) GetTwtTxnAfter(earliest int64, reqAddrNonce map[common.Uint160]uint64) ([]*pb.Transaction, error) {
+	respTxns := make([]*pb.Transaction, 0)
+
+	tp.twtMapMu.Lock()
+	defer tp.twtMapMu.Unlock()
+
+	for pair := tp.twtMap.Newest(); pair != nil; pair = pair.Prev() {
+
+		twt := pair.Value.(*txnWithTime)
+		if twt.arriveTime < earliest { // skip too old txn
+			break
+		}
+
+		sender, err := twt.txn.GetProgramHashes()
+		if err != nil {
+			return nil, err
+		}
+
+		addr := sender[0]
+		nonce, ok := reqAddrNonce[addr]
+
+		if ok { // if addr is in request list, we append greater nonce txns to response list
```

Comment:
We can simply it to `if !ok || twt.txn.UnsignedTx.Nonce < nonce`

Reply:
Yes, modified.